SDL/MiniDoublet.cu

```diff
@@ -147,448 +214,6 @@ void SDL::createMDsInExplicitMemory(struct miniDoublets& mdsInGPU, unsigned int
-    
-    mdsInGPU.moduleIndices[idx] = lowerModuleIdx;
-    unsigned int anchorHitIndex, outerHitIndex;
-    if(modulesInGPU.moduleType[lowerModuleIdx] == PS and modulesInGPU.moduleLayerType[lowerModuleIdx] == Strip)
-    {
-        mdsInGPU.anchorHitIndices[idx] = upperHitIdx;
-        mdsInGPU.outerHitIndices[idx] = lowerHitIdx;
-
-        anchorHitIndex = upperHitIdx;
-        outerHitIndex = lowerHitIdx;
-    }
-    else
-    {
-        mdsInGPU.anchorHitIndices[idx] = lowerHitIdx;
-        mdsInGPU.outerHitIndices[idx] = upperHitIdx;
-
-        anchorHitIndex = lowerHitIdx;
-        outerHitIndex = upperHitIdx;
-    }
-
-    mdsInGPU.dphichanges[idx] = dPhiChange;
-
-    mdsInGPU.dphis[idx] = dPhi;
-    mdsInGPU.dzs[idx] = dz;
-    mdsInGPU.shiftedXs[idx] = shiftedX;
-    mdsInGPU.shiftedYs[idx] = shiftedY;
-    mdsInGPU.shiftedZs[idx] = shiftedZ;
-
-    mdsInGPU.noShiftedDzs[idx] = noShiftedDz;
-    mdsInGPU.noShiftedDphis[idx] = noShiftedDphi;
-    mdsInGPU.noShiftedDphiChanges[idx] = noShiftedDPhiChange;
-//#ifdef CUT_VALUE_DEBUG
-//    mdsInGPU.dzCuts[idx] = dzCut;
-//    mdsInGPU.drtCuts[idx] = drtCut;
-//    mdsInGPU.miniCuts[idx] = miniCut;
-//#endif
-
-    mdsInGPU.anchorX[idx] = hitsInGPU.xs[anchorHitIndex];
-    mdsInGPU.anchorY[idx] = hitsInGPU.ys[anchorHitIndex];
-    mdsInGPU.anchorZ[idx] = hitsInGPU.zs[anchorHitIndex];
-    mdsInGPU.anchorRt[idx] = hitsInGPU.rts[anchorHitIndex];
-    mdsInGPU.anchorPhi[idx] = hitsInGPU.phis[anchorHitIndex];
-    mdsInGPU.anchorEta[idx] = hitsInGPU.etas[anchorHitIndex];
-    mdsInGPU.anchorHighEdgeX[idx] = hitsInGPU.highEdgeXs[anchorHitIndex];
-    mdsInGPU.anchorHighEdgeY[idx] = hitsInGPU.highEdgeYs[anchorHitIndex];
-    mdsInGPU.anchorLowEdgeX[idx] = hitsInGPU.lowEdgeXs[anchorHitIndex];
-    mdsInGPU.anchorLowEdgeY[idx] = hitsInGPU.lowEdgeYs[anchorHitIndex];
-
-    mdsInGPU.outerX[idx] = hitsInGPU.xs[outerHitIndex];
-    mdsInGPU.outerY[idx] = hitsInGPU.ys[outerHitIndex];
-    mdsInGPU.outerZ[idx] = hitsInGPU.zs[outerHitIndex];
-    mdsInGPU.outerRt[idx] = hitsInGPU.rts[outerHitIndex];
-    mdsInGPU.outerPhi[idx] = hitsInGPU.phis[outerHitIndex];
-    mdsInGPU.outerEta[idx] = hitsInGPU.etas[outerHitIndex];
-    mdsInGPU.outerHighEdgeX[idx] = hitsInGPU.highEdgeXs[outerHitIndex];
-    mdsInGPU.outerHighEdgeY[idx] = hitsInGPU.highEdgeYs[outerHitIndex];
-    mdsInGPU.outerLowEdgeX[idx] = hitsInGPU.lowEdgeXs[outerHitIndex];
-    mdsInGPU.outerLowEdgeY[idx] = hitsInGPU.lowEdgeYs[outerHitIndex];
-}
-
-ALPAKA_FN_ACC bool SDL::runMiniDoubletDefaultAlgoBarrel(struct modules& modulesInGPU, /*struct hits& hitsInGPU,*/ uint16_t& lowerModuleIndex, uint16_t& upperModuleIndex, unsigned int lowerHitIndex, unsigned int upperHitIndex, float& dz, float& dPhi, float& dPhiChange, float& shiftedX, float& shiftedY, float& shiftedZ, float& noshiftedDz, float& noShiftedDphi, float& noShiftedDphiChange, float xLower,float yLower, float zLower, float rtLower,float xUpper,float yUpper,float zUpper,float rtUpper)
```

Comment:
Any function that requires a templated input must be written directly in the header file to avoid linker errors. Also, any templated function that calls a device function, even if that device function does not have a templated input, also needs to be included directly in the header file. Here is Dan's explanation:

"The struct has to be instantiated somewhere that can see the code for the operator().  If the code is in the .cu file, then it can't be instantiated in Event.cu for the type of the template parameter TAcc.  Often for templated types that means including the code for the operator() fn in the header file.  Otherwise, the .cu file needs to instantiate it for every possible value of TAcc.  So try putting the code for the operator() inline in the header file, like in the alpaka example at https://alpaka.readthedocs.io/en/latest/basic/example.html


The technical term for this is "template specialization", where a template that is generic over the type TAcc gets specialized for a particular type of TAcc, the compiler has to be able to generate the code for that at the point that it needs to be specialized."

Reply:
Thanks a lot for the explanation!